### PR TITLE
feat: add `outdated` command

### LINKED ISSUE #6976: Add `outdated` command
Ala:

https://docs.npmjs.com/cli/v7/commands/npm-outdated

__The outdated command would only consider direct dependencies__.

It suggest that it would show the latest major, minor, and patch release.

### PR DESCRIPTION
**Pull request by sockmaster27**
Fixes #6976

### CODE REVIEW

**Comment by sockmaster27** [comment]
image

**Comment by sockmaster27** [comment]
image

**Comment by sockmaster27** [comment]
image

**Comment by sockmaster27** [comment]
Alternatively, no borders:

image

File: main/src/ca/uwaterloo/flix/api/Bootstrap.scala [hunk 1]
```diff
@@ -699,4 +701,92 @@ class Bootstrap(val projectPath: Path, apiKey: Option[String]) {
...
+          major.map(v => v.toString).getOrElse(""),
+          minor.map(v => v.toString).getOrElse(""),
+          patch.map(v => v.toString).getOrElse(""),
+        ))
+    }
+
+    if (rows.isEmpty) {
+      out.println(formatter.green(
+        s"""
+           |All dependencies are up to date
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/Bootstrap.scala:754]: Maybe in green? :D

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/Bootstrap.scala:754]: IT IS IN GREEN. I AM BLIND.

**Comment by sockmaster27** [comment]
> I see. What is the Flix object used for? Just for its formatter?

Yes, but it's pretty well in line with the rest of the bootstrap functions

**Comment by sockmaster27** [comment]
> I guess it does do dependency resolution first, but maybe that is OK, or even reasonable?

I think doing it any other way would be pretty annoying.

**Comment by magnus-madsen** [comment]
Roger. Ill have Matt take a peek or I will later.

File: main/src/ca/uwaterloo/flix/Main.scala [hunk 2]
```diff
@@ -421,6 +436,9 @@ object Main {
       cmd("release").text("  release a new version to GitHub.")
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/Main.scala:436]: releases

File: main/src/ca/uwaterloo/flix/Main.scala [hunk 3]
```diff
@@ -421,6 +436,9 @@ object Main {
...
         .action((_, c) => c.copy(command = Command.Release))
 
+      cmd("outdated").text("  show dependencies which have newer versions available.")
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/Main.scala:439]: shows

File: main/src/ca/uwaterloo/flix/api/Bootstrap.scala [hunk 4]
```diff
@@ -699,4 +701,92 @@ class Bootstrap(val projectPath: Path, apiKey: Option[String]) {
 
     Validation.success(())
   }
+
+  /**
+    * Show dependencies which have newer versions available.
+    */
+  def outdated(flix: Flix)(implicit out: PrintStream): Validation[Unit, BootstrapError] = {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/Bootstrap.scala:708]: There seems to be a lot going on here. A few suggestions:
- We can separate the "getting the data" part and the "formatting the data" part into their own functions.
- We might be able to clean up some of the match-on-Result stuff with a `for`.
- Comments describing what the major/minor/patch stuff does would help.
- Maybe major/minor/patch stuff could be part of the SemanticVersion namespace?

File: main/src/ca/uwaterloo/flix/util/Formatter.scala [hunk 5]
```diff
@@ -53,6 +53,27 @@ trait Formatter {
       leftline
   }
 
+  def table(colHeaders: List[String], colFormatters: List[String => String], rows: List[List[String]]): String = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/util/Formatter.scala:56]: Add comment. In particular about any invariants for the arguments.

File: main/src/ca/uwaterloo/flix/tools/pkg/SemVer.scala [hunk 6]
```diff
@@ -43,5 +43,55 @@ case class SemVer(major: Int, minor: Int, patch: Option[Int], buildDot: Option[I
...
 
+  /**
+    * Of the given `availableVersions`, get the newest version which is a major update, if one exists.
+    *
+    * Example:
+    *   - this = 1.2.0
+    *   - available versions = 1.2.0, 1.2.1, 1.3.0, 2.0.0
+    *   - -> 2.0.0
+    */
+  def majorUpdate(availableVersions: List[SemVer]): Option[SemVer] =
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/tools/pkg/SemVer.scala:54]: perhaps the argument name could just be versions

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Nice!

**Comment by sockmaster27** [comment]
@magnus-madsen 
Should we make it exit with a non-zero code to be used in CI and such?

**Comment by mlutze** [comment]
> @magnus-madsen 
> Should we make it exit with a non-zero code to be used in CI and such?

Yes please

**Pull request merged**